src/services/auth.service.ts

```diff
@@ -8,34 +8,119 @@ import AuthenticationError from "../common/utils/customErrors/authenticationErro
 import { redisClient } from "../storage/redisClient";
 import { authHelper } from "../common/helpers/auth.helper";
 import logger from "../common/logger/logger";
+import { signupDto } from "../data/dtos/signup.dto";
+import { Account } from "../models/account.model";
+import { hashHelper } from "../common/helpers/hash.helper";
+import { cityRepository } from "../repositories/city.repository";
+import clientError from "../common/utils/customErrors/clientError";
 
-class AuthService {
-  login = async (params: loginDto): Promise<string> => {
+export class AuthService {
+  isCityIdValid = async (id: string): Promise<boolean> => {
     try {
-      const account = await accountRepository.getUserAccount(params);
+      logger.info("is CityId Valid", {
+        cityId: id,
+        __filename,
+        functionName: "isCityIdValid",
+      });
+      const getCityId: string | null = await cityRepository.getCityId(id);
+      if (getCityId == null) {
+        throw new clientError("city id doesnot exist", id, 422);
+      } else {
+        return true;
+      }
+    } catch (err) {
+      throw err;
+    }
+  };
 
+  storeToken = async (token: string, userId: string): Promise<void> => {
+    try {
+      logger.info("store token in redis", {
+        userId,
+        __filename,
+        functionName: "storeToken",
+      });
+      await redisClient.setToken(userId, token); // set the JWT as the key and its value as valid
+      const payload = authHelper.verifyAccessToken(token); // verifies and decode the jwt to get the expiration date
+      const exp = payload.exp as number;
+      await redisClient.setExpireAt(userId, exp); // sets the token expiration date to be removed from the cache
+      logger.info("token storing successful", {
+        userId,
+        __filename,
+        functionName: "storeToken",
+      });
+    } catch (err) {
+      throw err;
+    }
+  };
+
+  registerUser = async (params: signupDto): Promise<string | null> => {
+    try {
+      logger.info("register user", {
+        query: params,
+        __filename,
+        functionName: "registerUser",
+      });
+      const account: Account | null = await accountRepository.getAccount(
+        params.username
+      );
+      if (account) {
+        throw new clientError(
+          "account with entered username already exists",
+          params.username,
+          409
+        );
+      }
+      //If city entered in sign up form then check if that city is in DB
+      if (params.cityId) {
+        await this.isCityIdValid(params.cityId);
+      }
+
+      const passwordHash: string = await hashHelper.generateHash(
+        params.password
+      );
+      const accountId: string = await accountRepository.createAccount(
+        params.username,
+        passwordHash
+      );
```

Comment:
@ShouryaMup-CAW Create the use first and use it's id to create a new account entry.

Reply:
We were told to create account first by Karthik, don't know the reason though